dnf/cli/commands/history.py

```diff
@@ -219,24 +243,140 @@ def str2transaction_id(s):
-        tids, merged_tids = self._args2transaction_ids()
+        environments = self.replay_data.get("environments", {})
+        if not isinstance(groups, dict):
+            raise dnf.exceptions.Error(
+                _('Unexpected type of "environments" in "{fn}", object expected.'.format(fn=fn))
+            )
 
+        self._na_reason_cache = {}
+
+        for na, pkg_data in rpms.items():
+            subj = hawkey.Subject(na)
+            nas = subj.get_nevra_possibilities(forms=[hawkey.FORM_NA])
+
+            if len(nas) != 1:
+                raise dnf.exceptions.Error(_('Cannot parse name.arch for package "{na}".').format(na=na))
+
+            parsed_na = nas[0]
+            try:
+                action = pkg_data["action"]
+
+                # validate the reasons before running the transaction and cache them
+                try:
+                    self._na_reason_cache[na] = \
+                        libdnf.transaction.StringToTransactionItemReason(pkg_data["reason"])
+                except IndexError as e:
+                    raise dnf.exceptions.Error(
+                        _('Unexpected value of package reason "{reason}" for package "{pkg}" in "{fn}".')
+                            .format(reason=pkg_data["reason"], pkg=na, fn=fn)
+                    )
+
+                if action in ("Install", "Upgrade", "Downgrade", "Reinstall"):
+                    nevra = "%s-%s.%s" % (parsed_na.name, pkg_data["to"], parsed_na.arch)
+                    if action in ("Upgrade", "Downgrade", "Reinstall"):
+                        from_nevra = "%s-%s.%s" % (parsed_na.name, pkg_data["from"], parsed_na.arch)
+                        query = dnf.subject.Subject(from_nevra).get_best_solution(self.base.sack)["query"]
+                        if not query.installed():
+                            raise dnf.exceptions.Error(
+                                _('Package "{pkg}" not found installed for action "{action}" in "{fn}".')
+                                    .format(pkg=from_nevra, action=action, fn=fn)
+                            )
+
+                elif action == "Removed":
+                    nevra = "%s-%s.%s" % (parsed_na.name, pkg_data["from"], parsed_na.arch)
+                else:
+                    raise dnf.exceptions.Error(
+                        _('Unexpected value of package action "{action}" for package "{pkg}" in "{fn}".')
+                            .format(action=action, pkg=na, fn=fn)
+                    )
+
+                query = dnf.subject.Subject(nevra).get_best_solution(self.base.sack)["query"]
+
+                if not query:
+                    raise dnf.exceptions.Error(_('Package "{nevra}" from "{fn}" was not found.')
+                        .format(nevra=nevra, fn=fn))
+
+                if action in ("Install", "Upgrade", "Downgrade"):
+                    sltr = dnf.selector.Selector(self.base.sack).set(pkg=query)
+                    self.base.goal.install(select=sltr, optional=False)
```

Comment:
I suggest `optional=not self.base.conf.strict`. The same construct is also used in other history transaction commands.
I also wonder why you prefer to not use base.install() method?

Reply:
> I suggest `optional=not self.base.conf.strict`. The same construct is also used in other history transaction commands.

As we've already discussed, I'm still firmly of the opinion the transaction needs to be strict here. I want to be sure the transaction is everything or nothing (unless we add a switch to loosen the strictness, but that should be specific to this command and not use the system default)

> I also wonder why you prefer to not use base.install() method?

I'm not entirely sure of this, it just seemed `base.install()` does a lot of unnecessary things I may not want here.

Reply:
Which ones?

Reply:
Not sure. It does something with `multilib_policy` and I'm not sure what the effect is there. Notably I now use  `with_provides=False, with_filenames=False` in `get_best_solution` as you suggested, `base.install()` doesn't do that, but that's probably minor. Should I use `base.install()` then?

Reply:
base.install() replaces part of your code. Also you can use `forms=` to define nevra form.

Reply:
I am now (in the new version) using `query.installed()` / `query.available()` so I don't think I can use the `base` methods.